creator-node/src/components/tracks/TranscodeDelegationManager.js

```diff
@@ -0,0 +1,436 @@
+      // do not pick self or a node that has already been chosen
+      if (currentSP === CREATOR_NODE_ENDPOINT || validSPs.has(currentSP)) {
+        continue
+      }
+      validSPs.add(currentSP)
+    }
+
+    return Array.from(validSPs)
+  }
+
+  static async handOffToSp(logger, { sp, req }) {
+    const { fileDir, fileName, fileNameNoExtension, uuid: requestID } = req
+
+    await TranscodeDelegationManager.fetchHealthCheck(sp)
+
+    logger.info({ sp }, `Sending off transcode and segmenting request...`)
+    const transcodeAndSegmentUUID =
+      await TranscodeDelegationManager.sendTranscodeAndSegmentRequest({
+        sp,
+        fileDir,
+        fileName,
+        fileNameNoExtension
+      })
+
+    return transcodeAndSegmentUUID
+  }
+
+  static async pollForTranscode(logger, { uuid, sp }) {
+    logger.info(
+      { sp },
+      `Polling for transcode and segments with uuid=${uuid}...`
+    )
+
+    return this.asyncRetry({
+      asyncFn: async (bail, num) => {
+        if (num === 50) {
+          bail(
+            new Error(`Transcode handoff max attempts reached for uuid=${uuid}`)
+          )
+          return
+        }
+
+        const { status, resp } =
+          await TranscodeDelegationManager.fetchTranscodeProcessingStatus({
+            sp,
+            uuid
+          })
+
+        // Have to throw errors to trigger a retry
+        if (!status) {
+          throw new Error(`Job for uuid=${uuid} has not begun yet...`)
+        }
+
+        if (status === 'IN_PROGRESS') {
+          throw new Error(`In progress for uuid=${uuid}...`)
+        }
+
+        if (status === 'FAILED') {
+          bail(
+            new Error(`Transcode handoff failed: uuid=${uuid}, error=${resp}`)
+          )
+          return
+        }
+
+        if (status === 'DONE') {
+          return resp
+        }
+      },
+      asyncFnTask: 'polling transcode',
+      retries: 50,
+      minTimeout: 10000,
+      maxTimeout: 10000
+    })
+  }
+
+  static async fetchFilesAndWriteToFs(
+    logger,
+    {
+      fileNameNoExtension,
+      transcodeFilePath,
+      segmentFileNames,
+      segmentFilePaths,
+      m3u8FilePath,
+      sp
+    }
+  ) {
+    let res
+
+    // TODO: parallelize?
+
+    // TODO: if any part of the below fails, we should remove any tmp dirs
+
+    // Get segments and write to tmp disk
+    logger.info({ sp }, `Fetching ${segmentFileNames.length} segments...`)
+    for (let i = 0; i < segmentFileNames.length; i++) {
+      const segmentFileName = segmentFileNames[i]
+      const segmentFilePath = segmentFilePaths[i]
+
+      res = await TranscodeDelegationManager.fetchSegment(
+        sp,
+        segmentFileName,
+        fileNameNoExtension
+      )
+      await Utils.writeStreamToFileSystem(res.data, segmentFilePath)
+    }
+
+    // Get transcode and write to tmp disk
+    logger.info({ sp, transcodeFilePath }, 'Fetching transcode...')
+    const transcodeFileName = fileNameNoExtension + '-dl.mp3'
```

Comment:
nit but can we have the filename management/creation live inside the `fetchTranscode`/`fetchM3u8File` methods themselves?